### PR TITLE
Add module to auto-write SQL Alchemy models and insert data

### PR DESCRIPTION
**Pull request by MaiMee1**
**status**: ready for review

Take in dataframe created from output from `read_xsv_file` (in `extended_csv.py`) and automatically write models (in python files) using the declarative method of SQL Alchemy. Shows interesting column stats like number of unique rows (to see if we should manually change it to a pk column or not) and the min and max value of length of string.

Bulk write models by directories. Bulk insert models by directories. Support TSV (untested) and CSV files. Tested with 10,000 rows taking some time.

### Effects
- `extended_csv.py`
  - Move functions `select_not_null` to `utils/dict_utils.py` for reuse
  - Correct type hint for files `TextIO` -> `IO`
- `utils.py`
  - Move functions to `utils/dict_utils.py` for grouping with new utils functions

**Requirements**
- pandas (new)
- stringcase

### Coverage
- Two test cases for auto-writing models
- Doctests in `str_utils.py` and `date_utils.py`

```
Name                                  Stmts   Miss Branch BrPart  Cover
-----------------------------------------------------------------------
sa_autowrite\__init__.py                  0      0      0      0   100%
sa_autowrite\hint.py                     20      4      0      0    80%
sa_autowrite\main.py                     92      6     30      3    93%
sa_autowrite\model.py                    91     14     28      5    84%
utils\__init__.py                         0      0      0      0   100%
utils\date_utils.py                      53      9     26      4    76%
utils\dict_utils.py                       7      1      8      0    80%
utils\io_utils.py                        14      2      4      2    78%
utils\str_utils.py                       23      2     14      0    95%
```

### CODE REVIEW

**Comment by MaiMee1** [comment]
Force push to remove wrongly committed temporary files 😅

File: tests/test_autowrite.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
+import unittest
+import pandas as pd
+
+
+TestCase = unittest.TestCase
+
+
+class TestAutowrite(TestCase):
+
+    def test_something(self):
```
Comment (th-bunratta) [inline tests/test_autowrite.py:10]: What's `test_something()`? More descriptive name?

**Review by th-bunratta — CHANGES_REQUESTED** [review]
Nothing much to say as some are imported modules. Keep up good work!

File: tests/test_autowrite.py (hunk 1 above)
Comment (MaiMee1) [reply tests/test_autowrite.py:10]: Great. Need to add tests for the new feature.

**Comment by MaiMee1** [comment]
Force pushed to take in test path bug fix (1645e0f) from master

File: tests/test_autowrite.py (hunk 1 above)
Comment (MaiMee1) [reply tests/test_autowrite.py]: Tests added

File: sa_autowrite/hint.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
+from abc import ABC, abstractmethod
+from typing import Dict
+
+from sqlalchemy.orm.attributes import InstrumentedAttribute
+from sqlalchemy.sql.schema import Table
+
+
+class DeclaredModel(ABC):
```
Comment (th-bunratta) [inline sa_autowrite/hint.py:8]: This class is used for type hinting, isn't it?

File: sa_autowrite/model.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
+from datetime import date, datetime
+from typing import List
+
+from pandas import DataFrame, Series
+
+from utils.date_utils import parse_date
+from utils.str_utils import camel_to_snake
+
+
+class Table:
```
Comment (th-bunratta) [inline sa_autowrite/model.py:10]: Is this an abstraction of the real SA Table implementation used in **autowrite** scripts?

**Review by th-bunratta — APPROVED** [review]
Looks good. Better merge the PR real soon to keep up with the pending works.

**Review by NutthanichN — APPROVED** [review]
I ran the sample code from `sa_autowrite/README.md` (after solved this [`numpy`'s `RuntimeError`](https://stackoverflow.com/questions/64729944/runtimeerror-the-current-numpy-installation-fails-to-pass-a-sanity-check-due-to)). It worked fine and shoud be ready to be merged.

**Comment by MaiMee1** [comment]
> I ran the sample code from `sa_autowrite/README.md` (after solved this [`numpy`'s `RuntimeError`](https://stackoverflow.com/questions/64729944/runtimeerror-the-current-numpy-installation-fails-to-pass-a-sanity-check-due-to)). It worked fine and shoud be ready to be merged.

😰 Seems like we should be careful about our environment and packages

File: sa_autowrite/hint.py (hunk 2 above)
Comment (MaiMee1) [reply sa_autowrite/hint.py:8]: Yes. Wanted some hints for the code, but not sure if it's correct or not. Shouldn't affect the program adversely tho.

File: sa_autowrite/model.py (hunk 3 above)
Comment (MaiMee1) [reply sa_autowrite/model.py:10]: More like a store of the data needed to write one such model. Not really related to SA's table, but just using the same name.

**Comment by MaiMee1** [comment]
Thanks for reviewing. I'll merge it now and open a new PR for further improvements for dealing with large data.

**Pull request merged**